backend/lucene/src/main/java/org/hibernate/search/backend/lucene/cache/QueryCachingConfigurationContext.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.lucene.cache;
+
+import java.util.Optional;
+import org.apache.lucene.search.QueryCache;
+import org.apache.lucene.search.QueryCachingPolicy;
+import org.apache.lucene.util.Version;
+
+/**
+ * A context allowing the configuration of query caching in a Lucene backend.
+ */
+public interface QueryCachingConfigurationContext {
+
+	/**
+	 * @return The Lucene version in use in the configured backend.
+	 */
+	Version luceneVersion();
+
+	/**
+	 * @param cache The {@link QueryCache} to use when searching.
+	 */
+	void queryCache(QueryCache cache);
+
+	/**
+	 * @return The corresponding {@link QueryCache}, or {@link Optional#empty()} if it doesn't exist.
+	 *
+	 */
+	Optional<QueryCache> queryCache();
```

Comment:
I'm still not following why you need this here :) Implementors are expected to define the cache and policy: why would they need to read it?
If you need this for implementation purposes, then it has nothing to do here in the API. Just expose it in the implementation class, but not in the interface.

Reply:
In service it can load successive implementations recursively. While defining, it was already added, you can make a delegating implementation that covers the existing one. Thanks to these parameters, the programmer knows that the object is already defined and has the ability to cover it by covering some of its functions. 

In other library:
```java 
@MetaInfServices(QueryCachingConfigurer.class)
public class OtherQueryCachingConfigurer implements QueryCachingConfigurer {

    @Override
    public void configure(QueryCachingConfigurationContext context) {
        context.queryCache(new GlobalQueryCache(context.luceneVersion()));
        context.queryCachingPolicy(new GlobalQueryCachingPolicy(context.luceneVersion()));
    }
}
```
In custom project:
```java
@MetaInfServices(QueryCachingConfigurer.class)
public class MyAppQueryCachingConfigurer implements QueryCachingConfigurer {

    @Override
    public void configure(QueryCachingConfigurationContext context) {
        
        context.queryCache().ifPresentOrElse((previous) -> {
            context.queryCache(new WrappedQueryCache(context.luceneVersion()));
        }, () -> {
            context.queryCache(new MyQueryCache(context.luceneVersion()));
        });
    }    
}
```

Reply:
The main problem I see with this is that there is not guarantee about the order in which the configurers are applied, apart from "implicit ones are applied before the one specified in the configuration".

If you have multiple caching configurers defined through the ServiceLoader, they can be applied in a different order depending on the JARs they come from, or even depending on the number of registered beans (due to `HashMap` rebalancings).

So yes, I think it's better to remove this until we have an actual use case.

If someone wants to apply multiple query caches, they can write their own "composite" configurer, which would delegate to other configurers in a well-defined order. And if that's not enough, they can submit a ticket and we can think better how to solve their problem once we have more information about the use case :)

Reply:
I'll change it.